### PR TITLE
upstream(graphql): add mutation payload size limit

### LINKED ISSUE #8726: upstream(graphql): Add a mutation payload size
https://github.com/MystenLabs/sui/pull/18017

### PR DESCRIPTION
**Pull request by kodemartin**
# Description of change

This patch introduces a new limit specific to transaction execution paths (`{mutation.execute,query.DryRun}TransactionBlock`. This allows transaction payloads to exceed the existing query limit size (5KB) and applies an upper that complies with the max size allowed by the protocol.

This fixes issues with large transactions (typically `publish`) not being able to execute.

`query.moveViewCall` is not handled here.

## Links to any relevant issues

Fixes #8726 

## How the change has been tested

Describe the tests that you ran to verify your changes.

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [x] Patch-specific tests (correctness, functionality coverage)

### Infrastructure QA (only required for crates that are maintained by @iotaledger/infrastructure)

- [x] Verification of API backward compatibility.

### Release Notes

- [x] GraphQL: Adds a new `maxTxPayloadSize` configuration parameter under the `limits` section. This is the maximum size in bytes allowed for the `txBytes` and `signatures` payload in `mutation.executeTransactionBlock` and for `txBytes` in `query.dryRunTransactionBlock`. The default value is the max transaction size allowed by the protocol (128KB) plus the `Base64` encoding overhead (~1/3 of the size).

### CODE REVIEW

**Review by thibault-martinez — APPROVED** [review]
dev-tools change LGTM

File: crates/iota-graphql-rpc/src/extensions/query_limits_checker.rs [hunk 1]
```diff
@@ -258,6 +405,17 @@ impl<'a> LimitsTraversal<'a> {
         Ok(())
     }
 
+    /// Error returned if transaction payloads exceed limit. Also sets the
+    /// transaction payload budget to zero to indicate it has been spent
+    /// (This is done to prevent future checks for smaller arguments from
+    /// succeeding even though a previous larger argument has already
+    /// failed).
+    fn tx_payload_size_error(&mut self) -> ServerError {
+        self.tx_payload_budget = 0;
```
Comment (tomxey) [inline crates/iota-graphql-rpc/src/extensions/query_limits_checker.rs]: I think this is a bit unexpected effect for a function that generates an error message.
Could the caller handle it? It's already manipulating the value of `tx_payload_budget` so it wouldnt be strange if it would also assign `tx_payload_budget=0`

File: crates/iota-graphql-rpc/src/server/builder.rs [hunk 2]
```diff
@@ -1168,4 +1174,794 @@ pub mod tests {
         let resp = reqwest::get(&url_with_param).await.unwrap();
         assert_eq!(resp.status(), reqwest::StatusCode::GATEWAY_TIMEOUT);
     }
+
+    /// Execute a GraphQL request with `limits` in place, expecting an error to
+    /// be returned. Returns the list of errors returned.
```
Comment (tomxey) [inline crates/iota-graphql-rpc/src/server/builder.rs]: nit: the return value is a string, so I think we cannot call it a "list" here

File: crates/iota-graphql-rpc/tests/e2e_tests.rs [hunk 3]
```diff
@@ -1004,12 +1033,192 @@ mod tests {
...
         cluster
             .wait_for_checkpoint_catchup(0, Duration::from_secs(10))
             .await;
         test_health_check_impl().await;
     }
+
+    #[tokio::test]
+    #[serial]
+    async fn test_payload_total_exceeded() {
+        test_payload_total_exceeded_impl().await;
```
Comment (tomxey) [inline crates/iota-graphql-rpc/tests/e2e_tests.rs:1052]: we have a lot of those payload size tests, what would you say about moving them to some dedicated file in `iota-graphql-rpc/tests/` ?

File: crates/iota-graphql-rpc/src/extensions/query_limits_checker.rs [hunk 4]
```diff
@@ -258,6 +405,17 @@ impl<'a> LimitsTraversal<'a> {
...
+        self.reporter
+            .payload_size_error("Transaction payload too large")
```
Comment (tomxey) [inline crates/iota-graphql-rpc/src/extensions/query_limits_checker.rs:413]: Could we also say what was the size of transaction payload? So the user knows how much he exceeded the value.

File: crates/iota-graphql-rpc/tests/e2e_tests.rs [hunk 5]
```diff
@@ -1004,12 +1033,192 @@ mod tests {
...
 
         cluster
             .wait_for_checkpoint_catchup(0, Duration::from_secs(10))
             .await;
         test_health_check_impl().await;
     }
+
+    #[tokio::test]
+    #[serial]
+    async fn test_payload_total_exceeded() {
```
Comment (tomxey) [inline crates/iota-graphql-rpc/tests/e2e_tests.rs:1051]: Could we also test the edge case where user submits the maximum transaction size allowed by the protocol?

File: crates/iota-graphql-rpc/src/config.rs [hunk 6]
```diff
@@ -277,7 +282,23 @@ impl ServiceConfig {
...
+    /// (including the signatures) allowed by the protocol, plus the Base64
+    /// overhead (roughly 1/3 of the original string).
+    async fn max_transaction_payload_size(&self) -> u32 {
+        self.limits.max_tx_payload_size
+    }
+
+    /// The maximum bytes allowed for the read part of GraphQL queries.
+    ///
+    /// In case of mutations or `dryRunTransactionBlocks` the `txBytes` and
+    /// signatures are not included in this limit.
```
Comment (tomxey) [inline crates/iota-graphql-rpc/src/config.rs]: nit: `signatures`

File: crates/iota-graphql-rpc/tests/e2e_tests.rs (hunk 5 above)
Comment (kodemartin) [reply crates/iota-graphql-rpc/tests/e2e_tests.rs:1051]: What would be the added value? This is a configuration parameter and we thoroughly test queries that exceed it.

File: crates/iota-graphql-rpc/src/extensions/query_limits_checker.rs (hunk 4 above)
Comment (kodemartin) [reply crates/iota-graphql-rpc/src/extensions/query_limits_checker.rs:413]: No we can't. This is lazily evaluated, so we just know it exceeded the limit set by the service. Can be more explitit about it here though.

File: crates/iota-graphql-rpc/tests/e2e_tests.rs (hunk 3 above)
Comment (kodemartin) [reply crates/iota-graphql-rpc/tests/e2e_tests.rs:1052]: I think this warrants for a more general restructuring, distinguishing between different scopes of the `e2e_tests`.

For example the payload tests go together with the query tests.

Will create a new issue.

Comment (kodemartin) [reply crates/iota-graphql-rpc/tests/e2e_tests.rs:1052]: https://github.com/iotaledger/iota/issues/9196

File: crates/iota-graphql-rpc/src/extensions/query_limits_checker.rs (hunk 1 above)
Comment (kodemartin) [reply crates/iota-graphql-rpc/src/extensions/query_limits_checker.rs]: 9bafa6942b

File: crates/iota-graphql-rpc/src/server/builder.rs (hunk 2 above)
Comment (kodemartin) [reply crates/iota-graphql-rpc/src/server/builder.rs]: 9bafa6942b

File: crates/iota-graphql-rpc/src/extensions/query_limits_checker.rs (hunk 4 above)
Comment (kodemartin) [reply crates/iota-graphql-rpc/src/extensions/query_limits_checker.rs:413]: 9bafa6942b

File: crates/iota-graphql-rpc/src/config.rs (hunk 6 above)
Comment (kodemartin) [reply crates/iota-graphql-rpc/src/config.rs]: 9bafa6942b

File: crates/iota-graphql-rpc/tests/e2e_tests.rs (hunk 5 above)
Comment (kodemartin) [reply crates/iota-graphql-rpc/tests/e2e_tests.rs:1051]: The configuration parameter is by default set in a higher value than the expected max protocol size plus the Base64 overhead. This should give enough leeway to account for additional syntactic checks we employ here. These syntactic checks cause an intrinsic tolerance with respect to the exact max protocol value, but it is ok to be conservative here.

If it is found that a well-constructed transaction query that is exactly the max protocol size is cut off here we can revisit.

**Review by tomxey — APPROVED** [review]
lgtm

**Review by sergiupopescu199 — APPROVED** [review]
LGTM 🌸

- We need to update the `.snap` file.
- Tests error messages does not match making the assert fail
```rust
/// Error returned if transaction payloads exceed limit.
    fn tx_payload_size_error(&mut self) -> ServerError {
        self.reporter.payload_size_error(
            format!(
                "Transaction payload too large: exceeds {}",
                self.reporter.limits.max_tx_payload_size
            )
            .as_str(),
        )
    }
```
They all expest this `exceeds {}` but is not present"

```
 thread 'tests::test_payload_dry_run_exceeded' panicked at /actions-runner/_work/iota/iota/crates/iota-graphql-rpc/src/server/builder.rs:1264:9:
    assertion `left == right` failed
      left: "Transaction payload too large: exceeds 10. Requests are limited to 10 bytes or fewer on transaction payloads (all inputs to executeTransactionBlock or dryRunTransactionBlock) and the rest of the request (the query part) must be 400 bytes or fewer."
     right: "Transaction payload too large. Requests are limited to 10 bytes or fewer on transaction payloads (all inputs to executeTransactionBlock or dryRunTransactionBlock) and the rest of the request (the query part) must be 400 bytes or fewer."
```
